ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM/QD.hs

```diff
@@ -0,0 +1,875 @@
+                (metadata, Just exn)
+            Nothing         ->
+                (metadata, Nothing)
+
+  -- stop the GSM
+  _ <- lift . lift $ uninterruptibleCancel hGSM
+
+  let modelStateAfterActions = QD.underlyingState metadata
+
+  let noExn = case mbExn of
+          Nothing  -> QC.property ()
+          Just exn -> QC.counterexample (show exn) False
+
+  -- effectively add a 'ReadGsmState' to the end of the command list
+  let readFinalStateAction = ReadGsmState
+  (finalStateCheck, finalModelState) <-
+    takeActionInBoth ("Difference after final action "  <> show readFinalStateAction)
+                     modelStateAfterActions
+                     readFinalStateAction
+
+  watcherEvents <- lift . lift $ dumpEvents varEvents
+
+  pure $ QC.counterexample
+            (unlines
+               $ (:) "WATCHER"
+               $ map (("    " <>) . show)
+               $ watcherEvents
+            )
+       $ QC.counterexample (("ediff Initial Final " <>)
+                           . Pretty.render . TD.prettyEditExpr
+                           $ TD.ediff model finalModelState)
+       $ QC.tabulate
+             "Notables"
+             (case Set.toList $ mNotables finalModelState of
+                []       -> ["<none>"]
+                notables -> map show notables
+             )
+       noExn QC..&&. finalStateCheck
+
+---
+
+-- | The parameters needed by 'Model' and the 'StateModel' instance.
+data StaticParams = StaticParams {
+    pUpstreamPeerBound :: Maybe UpstreamPeer
+  , pInitialJudgement  :: LedgerStateJudgement
+  , pIsHaaSatisfied    :: Set.Set UpstreamPeer -> Bool
+  }
+
+-- | The model state
+data Model s = Model {
+    mCandidates :: Map.Map UpstreamPeer Candidate
+  ,
+    mClock      :: SI.Time
+  ,
+    mIdlers     :: Set.Set UpstreamPeer
+  ,
+    mNotables   :: Set.Set Notable
+  ,
+    mPrev       :: WhetherPrevTimePasses
+  ,
+    mSelection  :: Selection
+  ,
+    mState      :: ModelState
+  }
+  deriving (Show, Generic)
+  deriving anyclass (TD.ToExpr)
+
+-- | Initialise the 'Model' state from 'StaticParams'
+initModel :: StaticParams -> Model params
+initModel StaticParams{pIsHaaSatisfied, pInitialJudgement} = Model {
+    mCandidates = Map.empty
+  ,
+    mClock = SI.Time 0
+  ,
+    mIdlers = idlers
+  ,
+    mNotables = Set.empty
+  ,
+    mPrev = WhetherPrevTimePasses True
+  ,
+    mSelection = initialSelection pInitialJudgement
+  ,
+    mState = case pInitialJudgement of
+        TooOld
+          | pIsHaaSatisfied idlers -> ModelSyncing
+          | otherwise             -> ModelPreSyncing
+        YoungEnough               -> ModelCaughtUp (SI.Time (-10000))
+  }
+  where
+    idlers = Set.empty
+
+-- | The 'StaticParams' are supplied through reflection because this seems to be
+--   the cleanest way to pass static configuration needed by the 'initState' and
+--   'arbitraryAction' methods
+instance Reflection.Reifies params StaticParams => QD.StateModel (Model params) where
```

Comment:
Intuitively, it seems to me that we could use the simpler [`Given`](https://hackage.haskell.org/package/reflection-2.1.9/docs/Data-Reflection.html#g:2) here and get rid of the `params` parameter everywhere, as `StaticParams` is supposed to be constant for a single test (ie there should be never more than one `StaticParams` dictionary in scope).